Implement getSimpleSearchWithModalityAndBodyPartPaged

zjp opened the pull request:
Closes #3 

I'm open to feedback!

kirbyju commented:
Thanks for submitting this!  A couple of minor details.

1) Can you provide a sample query to demonstrate usage on this in the comments?  I think that might help folks since this is a little more complicated than most of the other functions.  I am probably doing something silly, but I can't figure out how to avoid errors if I wanted to select more than one value for a given criteria.  E.g if I wanted to query for patients with CR or DX modalities in a single function call, what does that look like?  

2) I've been trying to organize the code into sections based on the API they use or the functionality they provide.  Can you put this down with the rest of the Advanced API functions?  They're near the end.

zjp commented:
Good question! I couldn't see from the documentation page how to get more than one value in. I went to the SimpleSearch webpage, opened the network inspector, and clicked on more than one modality to see what showed up. I discovered those modalities are sent as different parameters, e.g.:

```
criteriaType0: 
ImageModalityCriteria
value0: 
MG
criteriaType1: 
ImageModalityCriteria
value1: 
CR
criteriaType2: 
ImageModalityCriteria
value2: 
CT```

So this requires a small refactor. I'll push a new patch soon.

kirbyju commented:
Cool, thanks.  One other detail is that the dataframe and csv format options are smashing the majority of the output into the second column.  It might be nice to break that out into more columns so it's easier to read.

zjp commented:
OK, working on the DF and CSV options, but the refactor to support multiple arguments is done I think. I also added a docstring and put the function down with the rest of the Advanced API functions. Also added format checking for fromDate and toDate, and sensible defaults for them.

zjp commented:
I'm actually not sure how to parse this cleanly. You probably have more experience with DataFrames/CSV than I do. Does anything stick out at you for how to massage this data? Sample return pretty-printed below. 

```
[
   {
      "criteria":"Collections",
      "values":[
         {
            "criteria":"4D-Lung",
            "count":"20"
         }
      ]
   },
   {
      "criteria":"Species",
      "values":[
         {
            "criteria":"337915000",
            "count":"20"
         }
      ]
   },
   {
      "criteria":"Image Modality",
      "values":[
         {
            "criteria":"CT",
            "count":"20"
         },
         {
            "criteria":"RTSTRUCT",
            "count":"20"
         }
      ]
   },
   {
      "criteria":"Anatomical Site",
      "values":[
         {
            "criteria":"LUNG",
            "count":"20"
         }
      ]
   },
   {
      "criteria":"Manufacturer",
      "values":[
         {
            "criteria":"ADAC",
            "count":"20"
         },
         {
            "criteria":"Varian Imaging Laboratories, Switzerland",
            "count":"20"
         }
      ]
   }
]
```

kirbyju commented:
Yeah, it's a bit tricky.  I'm wondering if DF/CSV even make any sense as outputs on this particular function. Maybe we should just remove "format" as a parameter and ditch the related lines to convert the JSON?  Did you have a specific use case for this function?  What do you plan to do with the return values from this function after you run it?

zjp commented:
That last push did not remove formatting, but fixed an oversight: species arguments are NPEX Concept IDs, not strings. The function needed to translate between user facing arguments ('human', 'mouse', 'dog') and the IDs.

kirbyju commented on src/tcia_utils/nbia.py:
```diff
@@ -910,6 +928,153 @@ def getDoiMetadata(doi, output, api_url = "", format = ""):
+    if minStudies:
+        setOptionValue(Criteria.NumStudies, minStudies)
+        criteriaTypeIndex += 1
+    if modalityAnded:
```
After further testing, this may not have anything to do with the modalityAnded thing.  Even running getSimpleSearchCriteriaValues() with no parameters I get significantly lower numbers in the results than I see in the GUI with no search filters selected.  You can reproduce by looking at the numbers next to each item in the Simple Search left side menu at https://nbia.cancerimagingarchive.net/nbia-search/.

kirbyju replied:
@zjp I'm not sure if this discrepancy in search results compared to the NBIA GUI was ever something you were able to reproduce.  Also, I think it needs some updates to use the logging approach instead of print messages.  Can you take a look and let me know when you think you've got those things sorted out?

zjp replied:
I updated the code to use the logging approach, but I'm having trouble querying that endpoint -- not the whole advanced API, I don't think, just the one; it times out -- so I'm not able to look at it for now.

zjp replied:
Perhaps it would be useful to do away with modalityAnded for now and follow up on a later ticket. If I do

`nbia.getSimpleSearchCriteriaValues(collections=["4D-Lung"])`

I do get the correct values: 1 collection, 20 subjects, 1 species, 2 modalities, 1 anatomical site, 1 manufacturer.

kirbyju replied:
4D-Lung was one of the ones that looked okay to me before.  I tried with no parameters so it would just pull back everything and let me quickly scan the patient counts for the various collections/modalities/etc.  The numbers in the GUI were different on a decent percentage, but not all.

zjp replied:
Do you have any examples? When I try to query with no parameters, I get a gateway timeout -- understandable, since that's a lot of data to request. But it means I can't just go down the whole list all at once.

zjp replied:
Huh. I honestly don't see how it could be a client-side issue, but I can see that the simple search GUI uses a different endpoint to display results (for paging), and that endpoint's code is substantially shorter than this one's. 

See what the GUI uses
https://github.com/CBIIT/NBIA-TCIA/blob/master/software/nbia-api/src/gov/nih/nci/nbia/restAPI/GetSimpleSearchWithModalityAndBodyPartPaged.java

vs what I query here
https://github.com/CBIIT/NBIA-TCIA/blob/master/software/nbia-api/src/gov/nih/nci/nbia/restAPI/GetSimpleSearchCriteriaValues.java

kirbyju replied:
Based on https://wiki.cancerimagingarchive.net/display/Public/NBIA+Advanced+REST+API+Guide#NBIAAdvancedRESTAPIGuide-ExtendedSimpleSearchwithModalityandBodyPartPagedAPI, I think https://github.com/CBIIT/NBIA-TCIA/blob/master/software/nbia-api/src/gov/nih/nci/nbia/restAPI/GetSimpleSearchWithModalityAndBodyPartPaged.java is what's retrieving the information that appears in the search results on the right side of that screenshot.  I am not sure what API is pulling the numbers that appear next to the search criteria on the left side menu, but I was thinking it might be [GetSimpleSearchCriteriaValues.java](https://github.com/CBIIT/NBIA-TCIA/blob/master/software/nbia-api/src/gov/nih/nci/nbia/restAPI/GetSimpleSearchCriteriaValues.java).   I've emailed the lead developer of the GUI to weigh in.

kirbyju replied:
Actually, after you pointed me to looking for clues in the Network tab in your other PR I gave that a try for investigating this one.  I don't see [GetSimpleSearchCriteriaValues.java](https://github.com/CBIIT/NBIA-TCIA/blob/master/software/nbia-api/src/gov/nih/nci/nbia/restAPI/GetSimpleSearchCriteriaValues.java) in the list but I do see the separate calls for modality w/ counts, body part w/ counts, etc (https://nbia.cancerimagingarchive.net/nbia-api/services/getBodyPartValuesAndCounts) so perhaps he's just running several separate calls to get the left menu numbers.

kirbyju replied:
He confirmed he doesn't use [GetSimpleSearchCriteriaValues.java](https://github.com/CBIIT/NBIA-TCIA/blob/master/software/nbia-api/src/gov/nih/nci/nbia/restAPI/GetSimpleSearchCriteriaValues.java) anymore.  

I will create a bug report to have someone investigate both the gateway timeout and the inaccurate results that it returns.

zjp replied:
If the intention was to deprecate that method then I can pivot to querying the endpoint that's in use (getSimpleSearchWithModalityAndBodyPartPaged). It returns all of the same information that I wanted, plus some data that can be ignored.

kirbyju reviewed (commented):
I probably need to spend some more time on this, but I think we need to figure out why the search results are inaccurate before I dig any further.  I'm not seeing anything obvious that would cause that so hopefully something sticks out to you!

zjp commented on src/tcia_utils/nbia.py:
```diff
@@ -910,6 +928,153 @@ def getDoiMetadata(doi, output, api_url = "", format = ""):
+    minStudies: int          -- The minimum number of studies a collection must have to be included in the results
+    manufacturers: list[str] -- Imaging device manufacturers, e.g. SIEMENS
+    bodyParts: list[str]     -- Body parts of interest, e.g. CHEST, ABDOMEN
+    fromDate: str            -- First cutoff date, in DD/MM/YYYY format. Defaults to 1 January 1900
```
I'm confused by this request. If I go to the SimpleSearch webpage, open developer tools, go to the network tab, enter in a date range, then apply it, then I can see a call made to getSimpleSearchWithModalityAndBodyPartPaged (gSSWMABPP) with the dates in DD/MM/YYYY format. 

I certainly can change this code so that the user enters dates in YYYY/MM/DD format, but the other way around is what I see your web GUI using, and for an endpoint that isn't the odd one out you specified. Is this conversion happening on the server side for gSSWMABPP?

zjp replied:
Well actually in the GUI widget it's MM/DD/YYYY, it's the request that's DD/MM/YYYY.

kirbyju replied:
Interesting, there is even more variability in handling dates across the different parts of NBIA than I realized.  But to clarify the reason for my request, when I started working on tcia-utils the first date-related inputs I encountered were https://wiki.cancerimagingarchive.net/display/Public/NBIA+Search+REST+API+Guide#NBIASearchRESTAPIGuide-NewPatientsInCollection and https://wiki.cancerimagingarchive.net/display/Public/NBIA+Search+REST+API+Guide#NBIASearchRESTAPIGuide-NewStudiesInPatientCollection which both used the YYYY/MM/DD.  Then later I went to add support for https://wiki.cancerimagingarchive.net/display/Public/NBIA+Search+REST+API+Guide#NBIASearchRESTAPIGuide-getUpdatedSeries but noticed the API expects MM/DD/YYYY for that endpoint, but decided to do that conversion behind the scenes so that all date-related inputs for tcia-utils would be YYYY/MM/DD.  That's the only reason I'm suggesting we continue to stick with that format for getSimpleSearchCriteriaValues.

zjp replied:
OK, I'm not attached to either format, I just wanted to ensure if we changed it we'd actually be changing it to the consistent one. :) I'll push in a sec.

zjp commented on src/tcia_utils/nbia.py:
```diff
@@ -910,6 +928,153 @@ def getDoiMetadata(doi, output, api_url = "", format = ""):
+        metadata.raise_for_status()
+
+        # check for empty results and format output
+        if metadata.text and metadata.text != "[]":
```
Empty strings are falsy in Python, so if metadata.text is "" this will short circuit to false. Both possibilities end up being checked here.

kirbyju replied:
Oh, cool.  Now you know how much of a python noob I am :)

zjp commented on src/tcia_utils/nbia.py:
```diff
@@ -910,6 +928,153 @@ def getDoiMetadata(doi, output, api_url = "", format = ""):
+
+####### getSimpleSearchCriteriaValues function
+# Takes the same parameters as the SimpleSearch GUI 
+# Using more parameters narrows the values received. 
```
I've changed the word values to the phrase 'number of subjects' here.

zjp commented:
I've updated the title of this PR and will work to switch the code over to that endpoint too.

zjp commented:
OK, the latest push implements getSimpleSearchWithModalityAndBodyPartPaged

kirbyju commented:
Something's still not right here.  If I use your example query in the GUI I am getting different numbers than this API call is returning: 

nbia.getSimpleSearchWithModalityAndBodyPartPaged(collections=["TCGA-UCEC", "4D-Lung"], modalities=["CT"])

```
'totalPatients': 61,
 'bodyParts': [{'value': 'LUNG', 'count': 20},
  {'value': 'ABDOMEN', 'count': 1},
  {'value': 'UTERUS', 'count': 38},
  {'value': 'NOT SPECIFIED', 'count': 2}],
 'modalities': [{'value': 'CT', 'count': 61}],
 'collections': [{'value': '4D-Lung', 'count': 20},
  {'value': 'TCGA-UCEC', 'count': 41}],
 'species': [{'value': '337915000', 'count': 61}],
```
![image](https://user-images.githubusercontent.com/2379527/228561947-ba376793-b3e5-4bcc-8dcd-f78ee55acc91.png)

zjp commented:
I think I've narrowed it down. Can you re-run your query with minStudies=0 and see if that gives you the correct result? It appears to give me correct results. If so I will push a change to make the default 0 instead of 1.

kirbyju commented:
Works on my end too!

zjp commented:
Great to hear! I just pushed the updated code that sets minStudies to 0 by default.
